packages/contracts/contracts/router/modules/exchanges/NFTXModule.sol

```diff
@@ -0,0 +1,221 @@
+
+contract NFTXModule is BaseExchangeModule {
+    // --- Fields ---
+
+    INFTXMarketplaceZap public constant NFTX_MARKETPLACE =
+        INFTXMarketplaceZap(0x0fc584529a2AEfA997697FAfAcbA5831faC0c22d);
+
+    bytes4 public constant ERC721_INTERFACE = 0x80ac58cd;
+    bytes4 public constant ERC1155_INTERFACE = 0xd9b67a26;
+
+    // --- Constructor ---
+
+    constructor(address owner, address router)
+        BaseModule(owner)
+        BaseExchangeModule(router)
+    {}
+
+    // --- Fallback ---
+
+    receive() external payable {}
+
+   
+    // --- Multiple ETH listings ---
+
+    function buyWithETH(
+        INFTXMarketplaceZap.BuyOrder[] calldata orders,
+        ETHListingParams calldata params,
+        Fee[] calldata fees
+    )
+        external
+        payable
+        nonReentrant
+        refundETHLeftover(params.refundTo)
+        chargeETHFees(fees, params.amount)
+    {
+        uint256 length = orders.length;
+        for (uint256 i = 0; i < length; ) {
+            INFTXMarketplaceZap.BuyOrder memory order = orders[i];
+            // Execute fill
+            _buy(
+                orders[i],
+                params.fillTo,
+                params.revertIfIncomplete,
+                order.price
+            );
+            
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+
+    // --- Internal ---
+
+    function _buy(
+        INFTXMarketplaceZap.BuyOrder calldata buyOrder,
+        address receiver,
+        bool revertIfIncomplete,
+        uint256 value
+    ) internal {
+        // Execute the fill
+        try NFTX_MARKETPLACE.buyAndRedeem{value: value}(
+                buyOrder.vaultId, 
+                buyOrder.amount,
+                buyOrder.specificIds,
+                buyOrder.path,
+                receiver
+            )
+        {} catch {
+            // Revert if specified
+            if (revertIfIncomplete) {
+                revert UnsuccessfulFill();
+            }
+        }
+    }
+
+    // --- Single ERC721 offer ---
+
+    function sell(
+        INFTXMarketplaceZap.SellOrder[] calldata orders,
+        OfferParams calldata params,
+        Fee[] calldata fees
+    ) external nonReentrant {
+
+        uint256 length = orders.length;
+        for (uint256 i = 0; i < length; ) {
+            // Execute fill
+            _sell(
+                orders[i],
+                params.fillTo,
+                params.revertIfIncomplete,
+                fees
+            );
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function _sell(
+        INFTXMarketplaceZap.SellOrder calldata sellOrder,
+        address receiver,
+        bool revertIfIncomplete,
+        Fee[] calldata fees
+    ) internal {
+
+        IERC165 collection = sellOrder.collection;
+
+        if (collection.supportsInterface(ERC721_INTERFACE)) {
+            _approveERC721IfNeeded(IERC721(address(collection)), address(NFTX_MARKETPLACE));
```

Comment:
I more meant to move all logic inside the if:
- approve
- mint and sell
- refund